### ISSUE TITLE
Expose additional metadata via .getCompletions()

### ISSUE THREAD

**Issue by cspotcode** [yargs/yargs#1210]
I'd like to integrate yargs completions with PowerShell.  However, in PowerShell, tab completions are more than a list of strings.  They can include an (optionally different) display label and a tooltip.  For example, here are tab completions for npm commands:

![pwsh-tab-completions](https://user-images.githubusercontent.com/376504/44942531-4a13d400-ad81-11e8-8407-20935d8ffc1c.gif)

Aliases are indicated (for example "run -> run-script") and a description of the command is shown beneath.  For npm scripts, the script's contents are displayed.

I'd like to implement a generic library that provides these rich completions for yargs.  However, yargs doesn't expose the necessary information in its API.  I propose that `yargs.getCompletion()` supports a "rich" mode that returns an array of objects with extra metadata.  For example:

```
yargs.getCompletion(['foo', 'bar'], {mode: 'rich'}, (completions) => {
  // ...
});
```

The callback is passed an array of objects:

```
[{
    text: 'init',
    description: 'Create an empty repo',
    usage: 'command_hierarchy.js init [dir]'
}, {
    text: 'remote',
    description: 'Manage set of tracked repos',
    usage: 'command_hierarchy.js remote <command>'
}]
```

This data structure can be piped directly to PowerShell as JSON, so the code shouldn't be too complex for yargs.  I'm imagining something like:

```
# output completions as JSON
command_hierarchy.js --get-yargs-completions=rich
```

Then I can write a library that allows users to do this:

```
var yargs = require('yargs');
yargs
  .commandDir('cmds')
  .demandCommand()
  .completion();
require('pwsh-yargs-completions')(yargs); // register "completions-pwsh" command to generate PowerShell profile script, similar to .completion()
yargs
  .help()
  .argv;
```

**Comment by philcockfield** [comment]
To clarify, is there no other way to extract meta-data details about the 'options' for a command in yargs?

**Comment by cspotcode** [comment]
@philcockfield: Yeah, exactly.  Yargs in general doesn't support that kind of reflection.  For example, it's not possible to query yargs for a list of sub-commands, since yargs doesn't even know about sub-commands until it invokes the parent's `builder()`, and it won't do that unless you invoke the parent.  I think this is a code smell that should probably be tackled separately.  Probably just tech debt due to organic growth of the API.

Additionally, yargs' built-in completion mechanism does not expose the metadata required to enable the behavior I'm describing.  For example, yargs' built-in completion mechanism does not return any context about the completions it suggests.  It doesn't say, for example, "this completion is the name of a command."  So there's no way to reuse yargs' built-in completion mechanism, supplementing with additional information as a post-process step.  We have no choice but to reimplement yargs' completion mechanism.  This isn't efficient, since yargs is *already* doing that work and already has access to all that information.

**Comment by philcockfield** [comment]
Thanks for clarifying @cspotcode
